Solutions/Corvus.Sql.Tenancy/Corvus/Sql/Tenancy/Internal/TenantSqlConnectionFactory.cs

```diff
@@ -0,0 +1,242 @@
+        }
+
+        /// <summary>
+        /// Get a Sql container for a tenant.
+        /// </summary>
+        /// <param name="tenant">The tenant for which to retrieve the container.</param>
+        /// <param name="connectionDefinition">The details of the connection to create.</param>
+        /// <returns>A connection instance for the tenant.</returns>
+        public Task<SqlConnection> GetSqlConnectionForTenantAsync(ITenant tenant, SqlConnectionDefinition connectionDefinition)
+        {
+            if (tenant is null)
+            {
+                throw new System.ArgumentNullException(nameof(tenant));
+            }
+
+            if (connectionDefinition is null)
+            {
+                throw new System.ArgumentNullException(nameof(connectionDefinition));
+            }
+
+            SqlConnectionDefinition tenantedSqlConnectionDefinition = GetContainerDefinitionForTenant(tenant, connectionDefinition);
+            return this.CreateSqlConnectionAsync(tenant, connectionDefinition, tenantedSqlConnectionDefinition);
+        }
+
+        /// <summary>
+        /// Create the repository instance.
+        /// </summary>
+        /// <param name="tenant">The tenant.</param>
+        /// <param name="tenantedSqlConnectionDefinition">The connection definition, adapted for the tenant.</param>
+        /// <param name="configuration">The Sql configuration.</param>
+        /// <returns>A <see cref="Task"/> with completes with the instance of the document repository for the tenant.</returns>
+        protected async Task<SqlConnection> CreateSqlConnectionAsync(ITenant tenant, SqlConnectionDefinition tenantedSqlConnectionDefinition, SqlConfiguration configuration)
+        {
+            if (tenant is null)
+            {
+                throw new System.ArgumentNullException(nameof(tenant));
+            }
+
+            if (tenantedSqlConnectionDefinition is null)
+            {
+                throw new System.ArgumentNullException(nameof(tenantedSqlConnectionDefinition));
+            }
+
+            if (configuration is null)
+            {
+                throw new System.ArgumentNullException(nameof(configuration));
+            }
+
+            configuration.Database = string.IsNullOrWhiteSpace(configuration.Database)
+                ? tenantedSqlConnectionDefinition.Database
+                : configuration.DisableTenantIdPrefix
+                    ? configuration.Database
+                    : BuildTenantSpecificDatabaseName(tenant, configuration.Database);
+
+            return await this.CreateSqlConnectionAsync(configuration).ConfigureAwait(false);
+        }
+
+        private static string BuildTenantSpecificDatabaseName(ITenant tenant, string database) => $"{tenant.Id.ToLowerInvariant()}-{database}";
+
+        private async Task<SqlConnection> CreateSqlConnectionAsync(SqlConfiguration configuration)
+        {
+            if (string.IsNullOrEmpty(configuration.ConnectionStringSecretName) &&
```

Comment:
I'm not changing the pattern in this PR - this is how this fallback config has been handled in the other cases, and if we want to go and change that again, I'm happy to open up a new issue for it.

This is actually the very case that meant I wanted config reviewed top to bottom, because this configuration all fails if you don't instantiate the factory before attempting to access the config.

And when the whole config review process was complete, this was still like this, and still has the same problem.

So please do open a new issue for this and we can do the config review again.

Reply:
Don't fully understand, but I've created an issue where we can take this further https://github.com/corvus-dotnet/Corvus.Tenancy/issues/66